### ISSUE TITLE
Slightly different gradients on GPU

### ISSUE THREAD

**Issue by bkj** [pytorch/pytorch#2778]
Hi All --

I noticed that if I compute the gradient of a loss w.r.t. the parameters of the network multiple times, I get slightly different results.  ** EDIT: I've only observed this happening on the GPU, and appears to disappear when I set `torch.backends.cudnn.enabled=False`  That obviously slows things down though, so ideally I'd like to use `cudnn` if possible. **

```
class Net(nn.Module):
    def __init__(self):
        super(Net2, self).__init__()
        self.conv1 = nn.Conv2d(3, 10, kernel_size=5)
        self.conv2 = nn.Conv2d(10, 10, kernel_size=5)
        self.fc1 = nn.Linear(10, 16)
        self.fc2 = nn.Linear(16, 10)
        
    def forward(self, x):
        x = self.conv1(x)
        x = self.conv2(x)
        x = x.mean(dim=-1).mean(dim=-1)
        x = self.fc1(x)
        x = self.fc2(x)
        return x

# init network + optimizer
net = Net().cuda()
X_train, y_train = X_train.cuda(), y_train.cuda()
opt = torch.optim.SGD(net.parameters(), lr=0.1)

# compute gradient w.r.t. parameters
opt.zero_grad()
loss = F.cross_entropy(net(X_train), y_train)
loss.backward()
a = [p.grad.data.clone() for p in opt.param_groups[0]['params']]

# compute gradient again
opt.zero_grad()
loss = F.cross_entropy(net(X_train), y_train)
loss.backward()
b = [p.grad.data.clone() for p in opt.param_groups[0]['params']]

[(aa == bb).all() for aa, bb in zip(a, b)]
# returns [True, True, False, True, True, True, True, True]

a[2][0][0] - b[2][0][0]
# returns:
# 1.00000e-11 *
#   0.0000 -2.9104  2.9104  0.0000 -2.9104
#  -2.9104  0.0000 -2.9104  2.9104  2.9104
#   0.0000  0.0000  0.0000 -2.9104  2.9104
#   0.0000  2.9104  0.0000  0.0000  0.0000
#  -2.9104  0.0000 -2.9104  2.9104  2.9104
# [torch.cuda.FloatTensor of size 5x5 (GPU 0)]
```

I've only observed this happening on the GPU, rather than the CPU.  Also, when I change the number of channels in `conv1` and `conv2`, often the behavior goes away -- eg if I increase `10` to `32` in the above example, the behavior goes away.  Anyone have any thoughts on why this might be happening and how I might be able to mitigate it?

Thanks
Ben

PS -- If you want to replicate the error, the dataset is as follows:
```
from keras.datasets import mnist

(X_train, y_train), _ = mnist.load_data()
X_train, y_train = X_train[:100], y_train[:100]

X_train = Variable(torch.FloatTensor(X_train.astype('float'))).cuda()
y_train = Variable(torch.LongTensor(y_train.astype('int'))).cuda()

# expand to 3 channels
X_train = X_train.view(X_train.size(0), 1, 28, 28)
X_train = X_train.expand(X_train.size(0), 3, 28, 28)

X_train /= 255
```

**Comment by soumith** [comment]
this is expected, because of CuDNN being non-deterministic in it's backward w.r.t. weights.

As you've already found out, if you want determinism in this case you can add this to the top of your script:

```python
torch.backends.cudnn.enabled=False
```

The CuDNN library has several algorithms to compute backward, one with FFT, one with gemm, etc. and only some of them are non-deterministic. This is why you are seeing that if you change the sizes slightly, the algorithm is becoming deterministic.
